### PR TITLE
Changes to grammar in spec to make it closer to the p4c implementation

### PR DESCRIPTION
**Pull request by jafingerhut**

### CODE REVIEW

**Comment by jafingerhut** [comment]
I believe that none of these changes to the grammar in the spec represent a change in the syntax or the semantics of the language at all.  They simply make the grammar.mdk file in the spec more similar to the grammar used in the open source p4c compiler.  See comments I have made on the grammar.mdk changes for some explanation.

File: p4-16/spec/P4-16-spec.mdk [hunk 1]
```diff
@@ -3025,15 +3025,17 @@ The grammar production rule for general expressions is as follows:
...
 expression
     : INTEGER
+    | DOTS                           // DOTS is ...
+    | STRING_LITERAL
     | TRUE
     | FALSE
-    | STRING_LITERAL
-    | nonTypeName
-    | dotPrefix nonTypeName
+    | prefixedNonTypeName
```
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk:3032]: The definition of `prefixedNonTypeName` is `(nonTypeName | dotPrefix nonTypeName)`, so this and the similar change in grammar.mdk are not a change in the syntax.  This changed version is what p4c uses.

File: p4-16/spec/grammar.mdk [hunk 2]
```diff
@@ -383,6 +385,13 @@ tupleKeysetExpression
     ;
 // END:tupleKeysetExpression
 
+// BEGIN:optTrailingComma
+optTrailingComma
+    : /* empty */
+    | ","
+    ;
+// END:optTrailingComma
+
```
Comment (jafingerhut) [inline p4-16/spec/grammar.mdk:394]: No change in the definition of optTrailingComma -- simply moving its definition earlier in grammar.mdk to match the order that p4c defines these non-terminals relative to each other, so `diff` gives fewer differences.

Comment (jafingerhut) [inline p4-16/spec/grammar.mdk:394]: No change in the definition of optTrailingComma -- simply moving its definition earlier in grammar.mdk to match the order that p4c defines these non-terminals relative to each other, so `diff` gives fewer differences.

File: p4-16/spec/grammar.mdk [hunk 3]
```diff
@@ -541,7 +550,7 @@ headerStackType
 
 // BEGIN:specializedType
 specializedType
-    : prefixedType "<" typeArgumentList ">"
+    : typeName "<" typeArgumentList ">"
```
Comment (jafingerhut) [inline p4-16/spec/grammar.mdk:553]: typeName's only production rule is to expand to prefixedType, so this is not a change in syntax.  The p4c grammar uses the version with typeName.

File: p4-16/spec/grammar.mdk [hunk 4]
```diff
@@ -632,7 +634,7 @@ realTypeArgumentList
 // BEGIN:typeDeclaration
 typeDeclaration
     : derivedTypeDeclaration
-    | typedefDeclaration
+    | typedefDeclaration ";"
```
Comment (jafingerhut) [inline p4-16/spec/grammar.mdk:637]: This and the next change result in no syntax change, and matches p4c's grammar.

File: p4-16/spec/grammar.mdk [hunk 5]
```diff
@@ -877,6 +879,12 @@ actionRef
     ;
 // END:actionRef
 
+// BEGIN:entry
+entry
+    : optCONST optEntryPriority keysetExpression ':' actionRef optAnnotations ';'
+    ;
+// END:entry
+
```
Comment (jafingerhut) [inline p4-16/spec/grammar.mdk:887]: No change in definition of `entry` -- simply moving it earlier to match order that p4c defines them in.

**Review by mihaibudiu — APPROVED** [review]

File: p4-16/spec/P4-16-spec.mdk [hunk 6]
```diff
@@ -3025,15 +3025,17 @@ The grammar production rule for general expressions is as follows:
...
     | expression '[' expression ']'
     | expression '[' expression ':' expression ']'
-    | '{' expressionList '}'
-    | '{' kvList '}'
+    | '{' expressionList optTrailingComma '}'
```
[the remark this answers is not in the document]
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:3035]: They are one of the few grammar rules that is currently duplicated in both P4-16-spec.mdk and grammar.mdk, which might be part of the reason -- not sure about that.

**Review by jnfoster — APPROVED** [review]
Looks good. Thanks for doing this!

**Pull request merged**